polygeist/lib/polygeist/Passes/ParallelLICM.cpp

```diff
@@ -29,297 +32,548 @@ namespace {
+        builder.create<arith::AddIOp>(loop.getLoc(), std::get<0>(pair),
+                                      std::get<2>(pair)),
         std::get<1>(pair));
     cond = cond ? static_cast<Value>(
-                      b.create<arith::AndIOp>(loop.getLoc(), cond, val))
+                      builder.create<arith::AndIOp>(loop.getLoc(), cond, val))
                 : val;
   }
+  return cond;
+}
 
-  auto ifOp = b.create<scf::IfOp>(loop.getLoc(), TypeRange(), cond);
-  loop->moveBefore(ifOp.thenYield());
+scf::IfOp SCFParallelGuardBuilder::createGuard() const {
+  return builder.create<scf::IfOp>(loop.getLoc(), TypeRange(),
+                                   createGuardExpr());
+}
 
-  for (Operation *op : opsToMove)
-    loop.moveOutOfLoop(op);
+void SCFParallelGuardBuilder::replaceUsesOfLoopReturnValues(
+    scf::IfOp &ifOp) const {
+  // TODO
+}
 
-  LLVM_DEBUG({
-    loop.print(llvm::dbgs() << "\nModified loop:\n");
-    llvm::dbgs() << "\n----------------\n";
-  });
+//===----------------------------------------------------------------------===//
+// AffineForLoopGuardBuilder
+//===----------------------------------------------------------------------===//
+
+void AffineForGuardBuilder::guardLoop() const {
+  AffineIfOp ifOp = createGuard();
+
+  bool yieldsResults = !loop.getResults().empty();
+  if (yieldsResults) {
+    ifOp.getThenBodyBuilder().create<AffineYieldOp>(loop.getLoc(),
+                                                    loop.getResults());
+    ifOp.getElseBodyBuilder().create<AffineYieldOp>(loop.getLoc(),
+                                                    loop.getIterOperands());
+  }
+
+  loop->moveBefore(ifOp.getThenBlock()->getTerminator());
+  replaceUsesOfLoopReturnValues(ifOp);
 }
 
-void moveParallelLoopInvariantCode(AffineParallelOp loop) {
-  SmallVector<Operation *, 8> opsToMove;
-  collectHoistableOperation(loop, opsToMove);
-  if (opsToMove.empty())
-    return;
+IntegerSet AffineForGuardBuilder::createGuardExpr() const {
+  SmallVector<AffineExpr, 2> exprs;
+  SmallVector<bool, 2> eqflags;
 
-  // Move all operations we found to be invariant outside of the loop.
-  OpBuilder b(loop);
+  const AffineMap lbMap = loop.getLowerBoundMap();
+  const AffineMap ubMap = loop.getUpperBoundMap();
 
-  // TODO properly fill exprs and eqflags
+  for (AffineExpr ub : ubMap.getResults()) {
+    SmallVector<AffineExpr, 4> symbols;
+    for (unsigned idx = 0; idx < ubMap.getNumSymbols(); ++idx)
+      symbols.push_back(
+          getAffineSymbolExpr(idx + lbMap.getNumSymbols(), loop.getContext()));
+
+    SmallVector<AffineExpr, 4> dims;
+    for (unsigned idx = 0; idx < ubMap.getNumDims(); ++idx)
+      dims.push_back(
+          getAffineDimExpr(idx + lbMap.getNumDims(), loop.getContext()));
+
+    ub = ub.replaceDimsAndSymbols(dims, symbols);
+
+    for (AffineExpr lb : lbMap.getResults()) {
+      // Bound is whether this expr >= 0, which since we want ub > lb,
+      // we rewrite as follows.
```

Comment:
This doesn't seem to be always true, e.g.,
lb = 0, ub = 10, so loop is entered, even if step is 11.
but ub - lb - step = 10-0-11 = -1 !>= 0

Reply:
fixed